python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -325,7 +332,26 @@ def from_numpy(
             if not arr.flags.c_contiguous:
                 # We only natively support C-contiguous ndarrays.
                 arr = np.ascontiguousarray(arr)
-            pa_dtype = pa.from_numpy_dtype(arr.dtype)
+            try:
+                pa_dtype = pa.from_numpy_dtype(arr.dtype)
+            except pa.ArrowNotImplementedError as e:
+                if "Unsupported numpy type 17" in str(e):
+                    # Cannot convert dict in numpy array to pyarrow struct.
+                    # Instead, this will fall back to using pandas, which is
+                    # slower and consumes more memory.
+                    column_info = (
+                        f" in column named '{column_name}'" if column_name else ""
+                    )
+                    logger.warning(
+                        "Could not construct Arrow block from numpy array; encountered "
+                        f"`dict` values{column_info}, which cannot be casted to an "
+                        "Arrow data type. Falling back to using pandas "
+                        "block type, which is slower and consumes more memory. "
+                        "For maximum performance, consider expanding out each "
+                        "key-value pair in the dict into its own column in "
+                        "order to use native Arrow block types."
+                    )
+                    raise e
```

Comment:
The warning message looks good. but 2 small issues:
1. this might be too verbose. it's hard to suppress the message to only print once across the cluster. but we should at least make it print only once per process (which seems still a bit too verbose).
2. Let's also print a similar message for other error types? AFAIK, None values will fall back to pandas as well.